### PR TITLE
[4.0] SILGen/IRGen/KeyPaths: Components for ObjC properties need to be identified by selector.

### PR DESCRIPTION
**Pull request by jckarter**
Explanation: A property imported from Objective-C, or marked in Swift with the `dynamic` keyword, doesn't have a vtable slot, so can't be identified that way. Use the ObjC selector as the unique identifier to ascribe equality to such components.

Scope: Required for keypaths referencing ObjC or dynamic properties to work correctly.

Issue: rdar://problem/31768669

Risk: Low, part of a new feature that should have little interaction with existing code

Testing: Swift CI, test cases from internal Apple adopters

### CODE REVIEW

**Comment by jckarter** [comment]
@swift-ci Please test

**Comment by jckarter** [comment]
@slavapestov This look OK for the 4.0 branch?

**Comment by jckarter** [comment]
@swift-ci Please test

**Comment by jckarter** [comment]
@swift-ci Please test

**Comment by jckarter** [comment]
@swift-ci Please test

**Comment by jckarter** [comment]
@swift-ci Please test

File: lib/SILGen/SILGenExpr.cpp [hunk 1]
```diff
@@ -2703,7 +2703,12 @@ getIdForKeyPathComponentComputedProperty(SILGenFunction &SGF,
   }
   case AccessStrategy::DispatchToAccessor: {
     // Identify the property by its vtable or wtable slot.
-    return SILDeclRef(property->getGetter(), SILDeclRef::Kind::Func);
+    // Use the foreign selector if the decl is ObjC-imported, dynamic, or
+    // otherwise requires objc_msgSend for its ABI.
+    return SILDeclRef(property->getGetter(), SILDeclRef::Kind::Func,
```
Comment (slavapestov) [inline lib/SILGen/SILGenExpr.cpp]: Small cleanup suggestion for master: use the constructor that just takes a decl and a kind, and call asForeign() on it.

**Comment by slavapestov** [comment]
LGTM.

I don't know how much @tkremenek cares, but I like to justify the risk assessment, eg here it would be "Low, the changes are in code implementing a new feature and do not affect existing programs" etc.

**Review by slavapestov — APPROVED** [review]

**Comment by tkremenek** [comment]
@swift-ci clean test macOS

**Comment by tkremenek** [comment]
There was a timeout on CI.  Re-running.

**Comment by jckarter** [comment]
The CI passed on a previous build. I had to push an update to fix a merge conflict in the CHANGELOG but that wouldn't change the behavior of any code.

**Comment by tkremenek** [comment]
@jckarter OK

**Comment by jckarter** [comment]
@tkremenek Looks like the build got killed again. Mind force-merging this?

**Pull request merged**